xul/bootstrap.js

```diff
@@ -0,0 +1,308 @@
+      this.removeButton();
+    }
+
+    if (Services.prefs.getBoolPref(branch + "showContext")) {
+      this.removeContext();
+    }
+
+    this.tabBrowser = null;
+    this.browserWindow = null;
+  },
+
+  observe: function(aSubject, aTopic, aData) {
+    if (aTopic != "nsPref:changed") return;
+    switch (aData) {
+      case 'showButton':
+        if (Services.prefs.getBoolPref(branch + "showButton")) {
+          this.addButton();
+        } else {
+          this.removeButton();
+        }
+        break;
+      case 'showContext':
+        if (Services.prefs.getBoolPref(branch + "showContext")) {
+          this.addContext();
+        } else {
+          this.removeContext();
+        }
+        break;
+    }
+  },
+
+  openExplorerOrSave: function(aUrl, aSave) {
+    let crx_url, filename;
+    if (apmo_pattern.test(aUrl)) {
+      crx_url = this.tabBrowser.getBrowserForTab(this.tabBrowser.selectedTab).
+                  contentDocument.getElementsByClassName("dllink_green")[0].href;
+    } else if (apmo_download_pattern.test(aUrl)) {
+      crx_url = aUrl;
+    } else {
+      crx_url = get_crx_url(aUrl);
+    }
+    if (!crx_url) {
+      Cu.reportError('Cannot find extension URL');
+      return;
+    }
+    let match1 = apmo_download_pattern.exec(crx_url);
+    if (match1) {
+      let match2 = apmo_pattern.exec(this.tabBrowser.selectedTab.linkedBrowser.currentURI.spec);
+      if (match2) {
+        filename = match2[1] + "-" + match1[2] + ".zip";
+      } else {
+        filename = match1[1] + "-" + match1[2] + ".zip";
+      }
+    } else {
+      filename = get_zip_name(crx_url);
+    }
+    if (aSave) {
+      this.browserWindow.internalSave(crx_url, null, filename, null, "application/zip",
+        true, null, null, this.tabBrowser.selectedTab.linkedBrowser.currentURI,
+        this.tabBrowser.ownerDocument, false, null);
+    } else {
+      let newtab = this.tabBrowser.addTab("chrome://esrc-explorer/content/crxviewer.html" +
+                   '?' + encodeQueryString({crx: crx_url, zipname: filename}), {relatedToCurrent: true})
+      this.tabBrowser.moveTabTo(newtab, this.tabBrowser.tabContainer.selectedIndex + 1);
+      this.tabBrowser.selectedTab = newtab;
+    }
+  },
+
+  updateButton: function(aURI) {
+    let isExtUrl = cws_pattern.test(aURI.spec) || ows_pattern.test(aURI.spec) ||
+                   amo_pattern.test(aURI.spec) || amo_file_version_pattern.test(aURI.spec) ||
+                   apmo_pattern.test(aURI.spec);
+    if (isExtUrl) {
+      this.button.style.display = "block";
+    } else {
+      this.button.style.display = "none";
+    }
+  },
+
+  onLocationChange: function(aWebProgress, aRequest, aLocation, aFlag) {
+    this.updateButton(aLocation);
+  },
+
+  onClickButton: function(aEvent) {
+    this.openExplorerOrSave(this.tabBrowser.selectedTab.linkedBrowser.currentURI.spec, 
+                            aEvent.ctrlKey || aEvent.metaKey);
+  },
+
+  addButton: function() {
```

Comment:
Is it necessary to add `if (this.button) return;`, or otherwise assert that `this.button` is unset?